cli-plugins/plugin/plugin.go

```diff
@@ -0,0 +1,96 @@
+package plugin
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/docker/cli/cli"
+	cliplugins "github.com/docker/cli/cli-plugins"
+	"github.com/docker/cli/cli/command"
+	cliconfig "github.com/docker/cli/cli/config"
+	cliflags "github.com/docker/cli/cli/flags"
+	"github.com/docker/docker/pkg/term"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+// Run is the top-level entry point to the CLI plugin framework. It should be called from your plugin's `main()` function.
+func Run(makeCmd func(command.Cli) *cobra.Command, meta cliplugins.Metadata) {
+	// Set terminal emulation based on platform as required.
+	stdin, stdout, stderr := term.StdStreams()
+	logrus.SetOutput(stderr)
+
+	dockerCli := command.NewDockerCli(stdin, stdout, stderr, nil)
+
+	plugin := makeCmd(dockerCli)
+
+	cmd := newPluginCommand(dockerCli, plugin, meta)
+
+	if err := cmd.Execute(); err != nil {
+		if sterr, ok := err.(cli.StatusError); ok {
+			if sterr.Status != "" {
+				fmt.Fprintln(stderr, sterr.Status)
+			}
+			// StatusError should only be used for errors, and all errors should
+			// have a non-zero exit status, so never exit with 0
+			if sterr.StatusCode == 0 {
+				os.Exit(1)
+			}
+			os.Exit(sterr.StatusCode)
+		}
+		fmt.Fprintln(stderr, err)
+		os.Exit(1)
+	}
+}
+
+func newPluginCommand(dockerCli *command.DockerCli, plugin *cobra.Command, meta cliplugins.Metadata) *cobra.Command {
+	opts := cliflags.NewClientOptions()
+	var flags *pflag.FlagSet
+
+	name := plugin.Use
+	fullname := cliplugins.NamePrefix + name
+
+	cmd := &cobra.Command{
+		Use:              "docker" + " [OPTIONS] " + name + " [ARG...]",
+		Short:            fullname + " is a Docker CLI plugin",
+		SilenceUsage:     true,
+		SilenceErrors:    true,
+		TraverseChildren: false,
```

Comment:
When this is false it searches directly for the subcommand, so any plugin that has multiple layers of commands/subcommands, the parent flags will be ignored.
Example:

In my case normally, CLI command is
docker registry --host=1.1.1.1 <command>

which passes the host to all of the subcommands of the "registry" command.  Have verified that when this is false, it says the "host" flag is unrecognized, and it honors it when I changed this to "true".

Reply:
Thanks I think this is indeed wrong (or at least inconsistent wrt builtin type commands, which do use `true`). I'll flip this.